wurst/objects/items/ScrollsDefinition.wurst

```diff
@@ -0,0 +1,421 @@
+        ..setModelUsed(LocalItems.scrollHaste)
+        ..setInterfaceIcon(Icons.bTNScrollOfHaste)
+        ..setNameEnhance("Scroll of Haste")
+        ..setTooltipExtended(SH_TT)
+
+@compiletime function createTsunamiScroll() returns ItemDefinition
+    return createUsableItem(ITEM_SCROLL_TSUNAMI)
+        ..setCooldownGroup(ABILITY_TSUNAMI.toRawCode())
+        ..setLumberCost(9)
+        ..setAbilities(ABILITY_TSUNAMI.toRawCode())
+        ..setStockReplenishInterval(120)
+        ..setModelUsed(LocalItems.scrollPurple)
+        ..setInterfaceIcon(Icons.bTNSnazzyScrollPurple)
+        ..setNameEnhance("Scroll of Tsunami")
+        ..setTooltipExtended(ST_TT)
+
+@compiletime function createCycloneScroll() returns ItemDefinition
+    return createUsableItem(ITEM_SCROLL_CYCLONE)
+        ..setCooldownGroup(ABILITY_CYCLONE.toRawCode())
+        ..setLumberCost(9)
+        ..setAbilities(ABILITY_CYCLONE.toRawCode())
+        ..setStockReplenishInterval(120)
+        ..setModelUsed(LocalItems.scrollCyan)
+        ..setInterfaceIcon(Icons.bTNBansheeMaster)
+        ..setNameEnhance("Scroll of Cyclone")
+        ..setTooltipExtended(SC_TT)
+
+
+let BUFF_CYCLONE = 'Bcyc'
+let BUFF_CYCLONE_EXTRA = 'Bcy2'
+
+function onCyclone(unit target)
+    target.addAbility(GHOST_VIS_ID)
+    doPeriodicallyTimed(0.1, 15) buffPoller ->
+        if not target.hasAbility(BUFF_CYCLONE) and not target.hasAbility(BUFF_CYCLONE_EXTRA)
+            target.removeAbility(GHOST_VIS_ID)
+            buffPoller.stop()
+
+function unit.isFire() returns bool
+    return this.getTypeId() == UNIT_FIRE
+        or this.getTypeId() == UNIT_MAGE_FIRE_SUMMONED
+        or this.getTypeId() == UNIT_MAGE_FIRE
+
+function onTsunami(unit caster, vec2 target)
+    // The range from the caster that the effect is centered.
+    let OFFSET_DISTANCE = 300.
+
+    // The range from the center for which units are affected.
+    let EFFECT_DISTANCE = 150.
+
+    // Compute the center of the extra damage.
+    let center = caster.getPos().polarOffset(caster.getPos().angleTo(target), OFFSET_DISTANCE)
+
+    // Enumerate all units that are in region for extra damage.
+    ENUM_GROUP.enumUnitsInRange(center, EFFECT_DISTANCE)
+
+    // Perform the extra damage for the matching units.
+    for u in ENUM_GROUP
+        // Only enemy structures are affected.
+        if u.isEnemyOf(caster) and IsUnitType(u, UNIT_TYPE_STRUCTURE)
+            // Destroy fires outright.
+            if u.isFire()
+                flashEffect(Objects.nagaDeath, u.getPos())
+                u.kill()
+            // Otherwise deal additional damage.
+            else
+                caster.damageTarget(u, TSUNAMI_STRUCT_DMG)
+
+function onRoot(unit caster, unit target)
+    if target.isTroll()
+        InstantDummyCaster.castTarget(caster.getOwner(), ABILITY_ROOT_TROLL, 1, Orders.ensnare, target)
+    else
+        InstantDummyCaster.castTarget(caster.getOwner(), ABILITY_ROOT_HOSTILE, 1, Orders.entanglingroots, target)
```

Comment:
Any ability that does damage should not be cast using an instant dummy caster. The instant dummy caster can change owner immediately after it casts the spell, meaning the damage and the kill are made for the wrong player.

Reply:
Updated.